### ISSUE TITLE
Weird video decoding issue

### ISSUE THREAD

**Issue by qcasey** [Rafostar/clapper#10]
The floating mode seems really cool, looking to replace Celluloid with this! 

Unfortunately all media playback has this weird decoding issue, it looks neat but it's not intended I'm sure:

![Screenshot from 2020-11-18 08-31-26](https://user-images.githubusercontent.com/1429672/99558415-8e96e480-2978-11eb-935e-3281eb98c829.png)

Playing bbb_sunflower_1080p_30fps_normal.mp4 with the flatpak version, this is what's under the System tab:

GTK version: 3.99.4
GStreamer version: 1.18.1

Seemingly no important errors when running, either.

```bash
com.github.rafostar.Clapper

(com.github.rafostar.Clapper:2): Gtk-WARNING **: 08:28:58.999: Couldn't unexport handle for GdkX11Toplevel surface, unsupported windowing system
mesa: for the --simplifycfg-sink-common option: may only occur zero or one times!
mesa: for the --global-isel-abort option: may only occur zero or one times!
mesa: for the --amdgpu-atomic-optimizations option: may only occur zero or one times!
```

Arch Linux 5.4.77-1-lts Gnome 3.38.1 on X11. I know it's an early WIP so let me know if there's any other info you needed from me :)

**Comment by Rafostar** [comment]
> I know it's an early WIP so let me know if there's any other info you needed from me

First of all thank you for understanding that it is indeed using a lot of new/experimental features :smiley: 

> flatpak version, this is what's under the System tab:
>
> GTK version: 3.99.4
> GStreamer version: 1.18.1

If it is a Flatpak you do not have to report GTK and GStreamer versions, cause they are precompiled by me inside the package, so I know exactly what plugins and what versions are inside. So saying its Flatpak version will be enough.

As for output:
```sh
Couldn't unexport handle for GdkX11Toplevel surface, unsupported windowing system
```
This one is GTK4 bug on X11. It is related to file chooser and should not affect playback.

```sh
mesa: for the --simplifycfg-sink-common option: may only occur zero or one times!
mesa: for the --global-isel-abort option: may only occur zero or one times!
mesa: for the --amdgpu-atomic-optimizations option: may only occur zero or one times!
```
Those on the other hand are very important. They are related to `mesa` (your GPU drivers). The GPU seems to be having problems with VAAPI init.

Since this is on X11, I will start by asking if you enabled `vah264dec` in player prefs? I know that it improves performance, but it is experimental and seems to cause issues on X11, while working perfectly fine on Wayland. Make sure it is disabled (unticked). If it was enabled restart the player for the changes to take effect.

I think that it is probably the same issue as this one on mpv:
https://github.com/mpv-player/mpv/issues/7154

If `vah264dec` was disabled or disabling it does not help, try starting from command line like this:
```sh
EGL_PLATFORM=x11 GDK_BACKEND=x11 flatpak run com.github.rafostar.Clapper
```

**Comment by Rafostar** [comment]
Did some research. It seems that `EGL` support for x11 in GTK4 is not merged yet :disappointed:

GTK4 will try to use `GLX` on x11 and `EGL` on Wayland for now.

**Comment by qcasey** [comment]
> Did some research. It seems that EGL support for x11 in GTK4 is not merged yet

Thank you very much for researching this! I can confirm that switching over to wayland everything works like it should. Using 

`EGL_PLATFORM=x11 GDK_BACKEND=x11 flatpak run com.github.rafostar.Clapper`

Didn't make a difference. `vah264dec` is also not enabled. I'm not sure if those are still relevant with your GTK4 findings. It leads me to think my VAAPI driver may be misconfigured though, I'll look into that. 

Thanks again for your help! Will definitely return to this repo once I can fully switch to Wayland

**Comment by Rafostar** [comment]
@qcasey 

Thanks for verifying. On Wayland `vah264dec` can be enabled for reduced CPU usage.

From what I could gather, it looks more like GTK4 limitation to only support older `GLX` on x11, which unfortunately seems to be no longer compatible with newer AMD GPUs. But I do not have such new GPU, so I cannot confirm/test.

**Comment by qcasey** [comment]
Yes I agree, my RX 580 seems to be the limiting factor here. I switched over to the intel iGPU and Clapper works fine even in X11

**Comment by Rafostar** [comment]
It turns out to be caused by AMD driver defaulting to 10bit color depth. The older `gstreamer-vaapi` plugins assume the 8bit depth by default which leads to corruption. It also affects other applications.

According to [https://linuxreviews.org/Chromium](https://linuxreviews.org/Chromium)
```
This breaks video playback on AMD graphics cards with Mesa 20+ due to a known gstreamer-vaapi issue which will likely never be fixed.
```

There was an attempt to fix it in `GStreamer` some time ago, but it never succeeded:
[https://gitlab.freedesktop.org/gstreamer/gstreamer-vaapi/-/merge_requests/152](https://gitlab.freedesktop.org/gstreamer/gstreamer-vaapi/-/merge_requests/152)

And newer `va` series plugins do not work on GLX which is being used on X11. Wayland in GTK4 uses EGL (works much better/stable) and seems unaffected by this and also allows enabling `vah264dec` plugin that leads to even better performance.

Forcing 8bit color with Mesa env `allow_rgb10_configs=false` should fix the issue on X11.

### Long story short:
This can be fixed by starting Clapper like this:
```sh
allow_rgb10_configs=false flatpak run com.github.rafostar.Clapper
```
or by exporting `allow_rgb10_configs=false` in system environment (using `.drirc` will not work with Flatpak).

---

Still in case of `GTK4` and `GStreamer` Wayland works so much better and is preferred.
